pico_w/tls_client/picow_tls_client.c

```diff
@@ -0,0 +1,236 @@
+           so you may want to use a smaller fixed size buffer and copy the data to it using a loop, if memory is a concern */
+        char buf[p->tot_len + 1];
+
+        pbuf_copy_partial(p, buf, p->tot_len, 0);
+        buf[p->tot_len] = 0;
+
+        printf("***\nnew data received from server:\n***\n\n%s\n", buf);
+
+        altcp_recved(pcb, p->tot_len);
+    }
+    pbuf_free(p);
+
+    return ERR_OK;
+}
+
+static void tls_client_connect_to_server_ip(const ip_addr_t *ipaddr, TLS_CLIENT_T *state)
+{
+    err_t err;
+    u16_t port = 443;
+
+    printf("connecting to server IP %s port %d\n", ipaddr_ntoa(ipaddr), port);
+    err = altcp_connect(state->pcb, ipaddr, port, tls_client_connected);
+    if (err != ERR_OK)
+    {
+        fprintf(stderr, "error initiating connect, err=%d\n", err);
+        tls_client_close(state);
+    }
+}
+
+static void tls_client_dns_found(const char* hostname, const ip_addr_t *ipaddr, void *arg)
+{
+    if (ipaddr)
+    {
+        printf("DNS resolving complete\n");
+        tls_client_connect_to_server_ip(ipaddr, (TLS_CLIENT_T *) arg);
+    }
+    else
+    {
+        printf("error resolving hostname %s\n", hostname);
+        tls_client_close(arg);
+    }
+}
+
+
+static bool tls_client_open(const char *hostname, void *arg) {
+    err_t err;
+    ip_addr_t server_ip;
+    TLS_CLIENT_T *state = (TLS_CLIENT_T*)arg;
+
+    state->pcb = altcp_tls_new(tls_config, IPADDR_TYPE_ANY);
+    if (!state->pcb) {
+        printf("failed to create pcb\n");
+        return false;
+    }
+
+    altcp_arg(state->pcb, state);
+    altcp_poll(state->pcb, tls_client_poll, TLS_CLIENT_TIMEOUT_SECS * 2);
+    altcp_recv(state->pcb, tls_client_recv);
+    altcp_err(state->pcb, tls_client_err);
+
+    /* Set SNI */
+    mbedtls_ssl_set_hostname(altcp_tls_context(state->pcb), hostname);
+
+    printf("resolving %s\n", hostname);
+
+    // cyw43_arch_lwip_begin/end should be used around calls into lwIP to ensure correct locking.
+    // You can omit them if you are in a callback from lwIP. Note that when using pico_cyw_arch_poll
+    // these calls are a no-op and can be omitted, but it is a good practice to use them in
+    // case you switch the cyw43_arch type later.
+    cyw43_arch_lwip_begin();
+
+    err = dns_gethostbyname(hostname, &server_ip, tls_client_dns_found, state);
+    if (err == ERR_OK)
+    {
+        /* host is in DNS cache */
+        tls_client_connect_to_server_ip(&server_ip, state);
+    }
+    else if (err != ERR_INPROGRESS)
+    {
+        printf("error initiating DNS resolving, err=%d\n", err);
+        tls_client_close(state->pcb);
+    }
+
+    cyw43_arch_lwip_end();
+
+    return err == ERR_OK || err == ERR_INPROGRESS;
+}
+
+// Perform initialisation
+static TLS_CLIENT_T* tls_client_init(void) {
+    TLS_CLIENT_T *state = calloc(1, sizeof(TLS_CLIENT_T));
+    if (!state) {
+        printf("failed to allocate state\n");
+        return NULL;
+    }
+
+    return state;
+}
+
+void run_TLS_CLIENT_Test(void) {
```

Comment:
Looks a little bit weird that this function-name capitalises it as `TLS_CLIENT_`, whereas all the other function-names capitalise it as `tls_client_` ?
(although this is obviously only a very minor point)